// Copyright Â© 2017-2023 Trust Wallet.
//
// This file is part of Trust. The full Trust copyright notice, including
// terms governing use, modification, and redistribution, is contained in the
// file LICENSE at the root of the source code distribution tree.

syntax = "proto3";

package TW.InternetComputer.Proto;
option java_package = "wallet.core.jni.proto";

// Internet Computer Transactions
message Transaction {

    // Internet Computer Account
    message Account {
        string principal = 1;
        bytes subaccount = 2;
    }


    // Rename to ICP_Ledger_Send_Args
    // ICRC1 Ledger transfer arguments
    message ICRC1_Transfer_Args {
        bytes from_subaccount = 1;
        // TODO: change to string?
        Account to_account = 2;
        uint64 amount = 3;
        // TODO: fee is not needed
        uint64 fee = 4;
        uint64 memo = 5;
        // TODO: change field name to current_timestamp_secs
        uint64 created_at_time = 6;
    }

    // Payload transfer
    oneof transaction_oneof {
        ICRC1_Transfer_Args icrc1_transfer_args = 1;
    }
}

// TODO: Input data necessary to create a signed transaction.
message SigningInput {
    bytes private_key = 1;
    Transaction transaction = 2;
}

// Transaction signing output.
// TODO: Add a field for request result (success or failure)
// TODO: Add a field for original request message
// TODO: Add a field for signature components
// TODO: Add a field for error messages
message SigningOutput {
    // Signed and encoded transaction bytes.
    bytes encoded = 1;
}
